src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunLibraryRootProvider.cs

```diff
@@ -153,6 +171,15 @@ public static void CheckCanGenerateMethod(MethodDesc method)
             {
                 CheckTypeCanBeUsedInSignature(signature[i]);
             }
+
+            //
+            // CheckTypeCanBeUsedInSignature is insufficient - the ArgIterator used in GetCallRefMap() can be used to detect
+            // some cases (e.g. usage of types with indeterminate size) where compilation will eventually fail downstream.
+            //
+            // TODO: Is it possible to augment CheckTypeCanBeUsedInSignature() to accomplish the same? It is relative straightforward
+            // to check if the type has indeterminate size, but is that sufficient?
+            //
+            new GCRefMapBuilder(_targetDetails, false).GetCallRefMap(method);
```

Comment:
I think this is fine for now; if at some point we come to the conclusion this may be a perf issue, I believe we have two ways to proceed:

1) Split GetCallRefMap into two methods - a preparatory method including the checks and a secondary method producing the actual bit stream into the builder; it would be sufficient to call the preparatory method here;

2) Record the actual GC ref map produced by the builder and store it in the method instance so that we don't need to repeat the GC ref map emission for the method when building the executable.

Reply:
[Another advantage of the preparatory method would be that, without the actual emission of the GC ref map bitstream, we could avoid creating a new instance of the builder for each method. Admittedly it's a pretty small object so it should be relatively cheap.]

Reply:
Just had a quick chat with @SrivastavaAnubhav. According to his measurement, `GCRefMapNode.GetData()` is only 0.02% of the compilation time, I think we are fine here.